### PR TITLE
Adds a refund contract

### PR DESCRIPTION
**Pull request by alexvandesande**
This refund contract buys back tokens for how much it is able to secure and replaces them by a new token which can be used for future refunds.

### CODE REVIEW

File: refund.sol [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    // address of original DAO
+    token DAOTokens;
+    
+    // Address of new refund token (can be used to claim any additional eventual refund in the future)
+    token DAOBadgeOfHonor;
+    
+    // the price it will refund it for
+    uint public weiRefundedPerToken;
+
+    /* First time setup */
```
Comment (LefterisJP) [inline refund.sol]: Please provide some info here in the comments regarding the arguments. Especially some extra info aout the  BOHAddress which I assume is the 'Badge Of Honor' token address.

File: refund.sol [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    function DAORefund(address DAOAddress, address BOHAddress) {
+        DAOTokens = token(DAOAddress);
+        DAOBadgeOfHonor = token(BOHAddress);
+    }
+    
+    // to be called by the owner to lock the price so it guarantees all tokens are refunded the same
+    function setPrice() onlyOwner {
+        uint tokenSupply = DAOTokens.totalSupply() - DAOTokens.balanceOf(this);
```
Comment (LefterisJP) [inline refund.sol]: We will probably have to think a lot about the total supply here. Remember that the exploit allowed for people to create as many tokens as they want into a child DAO, stealing ether in the process and never burning any mother DAO tokens, thus not reducing the total supply of the mother DAO.

What about people who may have already gotten into a Dark DAO and are not malicious? We should probably add some kind of exception for them here.

I think that we should probably do some accounting, see who was naughty, see who may be an innocent victim in a dark DAO and then incorporate that accounting in here.

We should also consider blacklisting all known attackers from any refund. More money for the rest of the legitimate DTH.

File: refund.sol [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        weiRefundedPerToken = this.balance / tokenSupply;
+    }
+    
+    // allows tokens to be refunded by ether and badges of honors to be sent back
+    function convertTokens(uint amountOfTokens) {
+        if (weiRefundedPerToken == 0
+        || !DAOTokens.transferFrom(msg.sender, address(this), amountOfTokens))
```
Comment (LefterisJP) [inline refund.sol]: + 1 level of indentation here

File: refund.sol [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        if (weiRefundedPerToken == 0
+        || !DAOTokens.transferFrom(msg.sender, address(this), amountOfTokens)) 
+            throw;
+        
+        // do these last to avoid a recursion attack
+        DAOBadgeOfHonor.mintToken(msg.sender, amountOfTokens);
+        msg.sender.send(amountOfTokens * weiRefundedPerToken);
+    }
+
+    // Allows the refund contract on doing any actions except token transfers
```
Comment (LefterisJP) [inline refund.sol]: on doing -> to do

File: refund.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    function execute(address _target, bytes _data) onlyOwner  {
+        // Don't allow it to make a 'approve' (0x095ea7b3) action on the dao 
+        if (_target == address(DAOTokens)
```
Comment (LefterisJP) [inline refund.sol]: We need some more exceptions here. I made a PR to your PR: https://github.com/alexvandesande/DAO/pull/1

**Comment by CJentzsch** [comment]
cleaning up

**Pull request closed without merging**